[Relationships] Hasone and morphone

pxpm opened the pull request:
## WHY

Make `HasOne` and `MorphOne` relations consistent without any difference on the setup, cause there should be no difference, the only difference is at database level, and that should be handled by Laravel as part of the relations.

### BEFORE - What was wrong? What was happening before this PR?

`HasOne` was working almost fine, it was just impossible to clear a value from a `BelongsTo` relation **nested** in an `HasOne` and the field would not auto-set as a select for the relation. 
MorphOne partially worked, needed to add some more configs to the field, `'morphs' => true` but didn't support relations like HasOne supports. 

### AFTER - What is happening after this PR?

- Setting up` HasOne` or `MorphOne` relation in CrudPanel is exactly the same, Backpack takes care of saving the correct relation. 
- `BelongsTo` relations can be properlly cleared.


## HOW

### How did you achieve that, in technical terms?

I didn't pursue the `morphs` configuration scenario since `HasOne` was working in other part of the panel, and `MorphOne` should be exactly the same as `HasOne` in terms of saving process. I didn't remove the `morphs` part from `syncPivot()` function for Backwards compatibility, but we should be aware that people using the new relationships will never reach that part of the code. 



### Is it a breaking change or non-breaking change?

I don't think it's breaking, but to be safe this is pointing to 4.2


### How can we test the before & after?

This relationships are setup in branch 4.2 of demo project. Also there are tests in the test suit to test the saving process. 

There are failing scenarios that I will be adding into #4012

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -209,38 +190,35 @@ private function createRelationsForItem($item, $formattedData)
+            if (! array_key_exists('model', $fieldData)) {
+                $fieldData['model'] = $relation_field['model'];
             }
+            if (! array_key_exists('parent', $fieldData)) {
+                $fieldData['parent'] = $this->getRelationModel($attributeKey, -1);
+            }
```
```suggestion
            $fieldData['parent'] = $fieldData['parent'] ?? $this->getRelationModel($attributeKey, -1);
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -209,38 +190,35 @@ private function createRelationsForItem($item, $formattedData)
-                Arr::set($relationData, 'relations.'.$key, $fieldData);
+            $key = implode('.relations.', explode('.', $this->getOnlyRelationEntity($relation_field)));
+            $fieldData = Arr::get($relationDetails, 'relations.'.$key, []);
+            if (! array_key_exists('model', $fieldData)) {
+                $fieldData['model'] = $relation_field['model'];
             }
```
```suggestion
            $fieldData['model'] = $fieldData['model'] ?? $relation_field['model'];
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -209,38 +190,35 @@ private function createRelationsForItem($item, $formattedData)
 
-        $relationData = [];
+        $relationDetails = [];
         foreach ($relation_fields as $relation_field) {
```
Inside this `foreach` everything is a bit obscure... I can't easily tell what's happening. Can you please add comments explaining what is happening and why?

Thanks.

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -189,11 +189,40 @@ public function guessIfFieldHasPivotFromRelationType($relation_type)
+     *
+     * @return array The fields with n-n relationships.
+     */
+    public function getRelationFieldsWithPivot()
+    {
+        $all_relation_fields = $this->getRelationFields();
+
+        return Arr::where($all_relation_fields, function ($value, $key) {
+            return isset($value['pivot']) && $value['pivot'];
+        });
+    }
```
Good idea moving this here! 💪

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -145,52 +132,46 @@ public function syncPivot($model, $input)
+     * @param  array  $input  The form data.
      */
-    private function createOneToOneRelations($item, $data)
+    private function createOneToOneRelations($item, $input)
```
Yeeey!! One more `$data` gone 🎉🎉🎉

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -209,38 +190,35 @@ private function createRelationsForItem($item, $formattedData)
+     * @return array The formatted relation details.
      */
-    private function getRelationDataFromFormData($data)
+    private function getRelationDetailsFromInput($input)
```
Much better name! 🙏

tabacitu replied:
@pxpm it looks like this _always_ returns an array with one item ("relations"), that is an array that holds the relations. Am I mistaken? Any reason to do this?

```php
// instead of returning
[
    'relations' => [
         'comments' => [],
         'stars' => [],
    ]
]
// wouldn't it be simpler to return this?
[
     'comments' => [],
     'stars' => [],
]
```

If that's possible, I think this would simplify the `Arr:get()` and `Arr:set()` too, no more need for the `.relations.` prefix. That would make the code more readable.

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -145,52 +132,46 @@ public function syncPivot($model, $input)
-                    $relation->save($modelInstance);
-                }
+            } elseif ($relation instanceof HasOne || $relation instanceof MorphOne) {
+                $modelInstance = $relation->updateOrCreate([], $relationDetails['values']);
```
Great use of `updateOrCreate()` 👍  When I see 7 lines turning into 1 simple line... I know we've made progress 😀

tabacitu reviewed (changes requested):
Looking good! 💪  Love the new variable and method names. This PR will not only bring uniformity between `HasOne` and `MorphOne`, but will also tidy up this bit of code, which I love. Like on the mountain - "_Leave the campsite cleaner than you found it._" 🏕

There's a `foreach` inside `getRelationDetailsFromInput()` where voodoo is still happening, though 😅  Can we take the opportunity to make that part clear too? With more comments at least?

Thanks!

pxpm commented:
Hey @tabacitu 

I've refactored the conditionals and added some comments. 

Like I told you, we should and partially support nested relations inside hasone. 

There are some code I will be adding here in future to fully support nested relations in hasone and morphone. In a later PR. 

Let me know if my comments didn't hit the point, I can try to rephrase them.

Best,
Pedro

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -209,38 +190,42 @@ private function createRelationsForItem($item, $formattedData)
-                Arr::set($relationData, 'relations.'.$key, $fieldData);
-            }
+            // since we can have for example 3 fields for address relation, we make sure that atleast once we set the relation details.
+            $fieldDetails = Arr::get($relationDetails, 'relations.'.$key, []);
+
+            $fieldDetails['model'] = $fieldDetails['model'] ?? $relation_field['model'];
+
+            $fieldDetails['parent'] = $fieldDetails['parent'] ?? $this->getRelationModel($attributeKey, -1);
```
```suggestion
            $fieldDetails = Arr::get($relationDetails, 'relations.'.$key, []);
            $fieldDetails['model'] = $fieldDetails['model'] ?? $relation_field['model'];
            $fieldDetails['parent'] = $fieldDetails['parent'] ?? $this->getRelationModel($attributeKey, -1);
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -209,38 +190,42 @@ private function createRelationsForItem($item, $formattedData)
 
-                Arr::set($relationData, 'relations.'.$key, $fieldData);
-            }
+            // since we can have for example 3 fields for address relation, we make sure that atleast once we set the relation details.
```
```suggestion
            // since we can have for example 3 fields for address relation, we make sure that at least once we set the relation details.
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -57,16 +57,16 @@ public function inferRelationTypeFromRelationship($field)
     }
 
-    public function getOnlyRelationEntity($relation_field)
+    public function getOnlyRelationEntity($field)
```
@pxpm heads-up: I've taken the liberty to rename the variables here and use the Fluent string helper, no changes in behavior.

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -145,102 +133,97 @@ public function syncPivot($model, $input)
+            // (user -> HasOne accountDetails -> BelongsTo address)
+            // we specifically use only the relation entity because relations like
+            // HasOne and MorphOne use the attribute in the relation string
+            $key = implode('.relations.', explode('.', $this->getOnlyRelationEntity($field)));
```
Please see my comment above. Any chance we can remove this exploding/imploding with `.relations.` by moving everything in the resulting array up one level? Looks very intimidating - is it really needed?

pxpm replied:
IT create a `relations` for relations inside relations too:

```php
[
    'relations' => [
        'some_relation' => [
           'relations' => [
           ] 
       ]
    ]
]
```
In the example in the comments it would return `relations => HasOne => relations => belongsto`

tabacitu replied:
Ooooh.... ok... hmm... Ok then, maybe we find a clearer way another day, problem for future us. It's a `private` method after all. 

Merging as-is, thank you 🙏

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -145,102 +133,97 @@ public function syncPivot($model, $input)
+
+            // since we can have for example 3 fields for address relation,
+            // we make sure that at least once we set the relation details
+            $fieldDetails = Arr::get($relationDetails, 'relations.'.$key, []);
```
If we move everything up one level, wouldn't this do the same as:

```suggestion
            $fieldDetails = $relationDetails[$key] ?? [];
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -145,102 +133,97 @@ public function syncPivot($model, $input)
+            $fieldDetails['parent'] = $fieldDetails['parent'] ?? $this->getRelationModel($field['name'], -1);
+            $fieldDetails['values'][$attributeName] = Arr::get($input, $field['name']);
+
+            Arr::set($relationDetails, 'relations.'.$key, $fieldDetails);
```
Here too - if we move everything up one level, wouldn't this do the same?

```suggestion
            $relationDetails[$key] = $fieldDetails;
```

or maybe
```suggestion
            $relationDetails[$key] = $relationDetails[$key] ?? $fieldDetails;
```

tabacitu reviewed (changes requested):
Thanks a lot @pxpm . I've taken one pass at it too, in trying to understand what that `foreach` does. The comments help a lot, but I still think we can do one more thing - move everything up one level. From what I see, this method always returns just one thing, an array with one `relations` item. I don't see a good reason for that, so let's take this opportunity to fix that too, if possible. Thank you 🙏
